eventstream/memorystream/stream_test.go

```diff
@@ -35,20 +38,168 @@ var _ = Describe("type Stream", func() {
 })
 
 var _ = Describe("type Stream", func() {
-	Describe("func Append()", func() {
-		It("panics if the message type is not supported", func() {
-			p := NewParcel("<id>", MessageA1)
+	var (
+		ctx    context.Context
+		cancel context.CancelFunc
+		stream *Stream
+	)
 
-			stream := &Stream{
-				App: configkit.MustNewIdentity("<app-name>", "<app-key>"),
-				Types: message.NewTypeSet(
-					MessageBType,
-				),
-			}
+	BeforeEach(func() {
+		ctx, cancel = context.WithTimeout(context.Background(), 1*time.Second)
+
+		stream = &Stream{
+			App: configkit.MustNewIdentity("<app-key>", "<app-name>"),
+			Types: message.NewTypeSet(
+				MessageEType,
+			),
+			// For the purposes of our test, we assume there are already 100
+			// events in the store.
+			FirstOffset: 100,
+		}
+	})
+
+	AfterEach(func() {
+		cancel()
+	})
+
+	Describe("func Add()", func() {
+		When("the events are the next expected events", func() {
+			It("makes the events available immediately", func() {
+				addEvents(stream, 100, 101)
+				expectEventsToBeAvailable(ctx, stream, 100, 101)
+			})
+		})
+
+		When("the events are added out of order", func() {
+			It("does not skip past the 'missing' events", func() {
+				addEvents(stream, 102, 103)
+
+				ctx, cancel := context.WithTimeout(ctx, 20*time.Millisecond)
+				defer cancel()
+
+				cur, err := stream.Open(ctx, 100, stream.Types)
+				Expect(err).ShouldNot(HaveOccurred())
+				defer cur.Close()
+
+				_, err = cur.Next(ctx)
+				Expect(err).To(Equal(context.DeadlineExceeded))
+			})
 
-			Expect(func() {
-				stream.Append(p)
-			}).To(Panic())
+			It("reorders the events", func() {
+				addEvents(stream, 102, 103)
+				addEvents(stream, 104, 105)
+				addEvents(stream, 100, 101)
+				expectEventsToBeAvailable(ctx, stream, 100, 105)
+			})
+		})
+
+		When("all of the events are older than the oldest events already in the stream", func() {
+			It("discards the events", func() {
+				addEvents(stream, 100, 101)
+				addEvents(stream, 98, 99)
+
+				expectEventsToBeTruncated(ctx, stream, 98, 99)
+				expectEventsToBeAvailable(ctx, stream, 100, 101)
+			})
+		})
+
+		When("some of the events are older than the oldest events already in the stream", func() {
+			It("discards only the older events", func() {
+				addEvents(stream, 99, 101)
+
+				expectEventsToBeAvailable(ctx, stream, 100, 101)
+			})
+		})
+
+		When("the maximum buffer size is exceeded", func() {
+			It("drops old events when the max buffer size is reached", func() {
+				stream.BufferSize = 3
+
+				addEvents(stream, 100, 101)
+				addEvents(stream, 102, 103)
+
+				expectEventsToBeTruncated(ctx, stream, 100, 100)
+				expectEventsToBeAvailable(ctx, stream, 102, 103)
```

Comment:
This isn't wrong, but should this also be expecting `101` to be available? A left over from your old batch version of this PR?